src/Microsoft.ML.Data/TrainCatalog.cs

```diff
@@ -20,7 +19,8 @@ namespace Microsoft.ML
     /// </summary>
     public abstract class TrainCatalogBase
     {
-        protected internal readonly IHost Host;
+        [BestFriend]
+        private protected readonly IHost Host;
```

Comment:
>private [](start = 8, length = 7)

is this attribute valid on a private field? ("A private protected member is accessible by types derived from the containing class, but only within its containing assembly.")

Reply:
This is not a private field. See the modifier `private protected`.

Reply:
This field seems to be also used in RecommenderCatalog.cs  (which is in a separate assembly Microsoft.ML.Recommender).  

So the [BestFriend] attribute is valid  for both `internal` and `private protected` access modifiers. Am i right ?

Reply:
Yes, this is a very quirky behaviour of C# 7.2 Apparently they treat private protected as "internal" from the point of view of the InternalsVisibleTo attribute.  This is actual behaviour and under the hood the CLR marks the private protected as protectedAndInternal accesibility level.  Very confusing. But the bottom line is that "private protected" field is only visible to derived classes in the same assembly and any assembly that is a friend to it.

---
In reply to: [256224446](https://github.com/dotnet/machinelearning/pull/2526#discussion_r256224446) [](ancestors = 256224446)

Reply:
I understand, but it is confusing. Perhaps we can update the BestFriend attribute xml docs to say 
    /// <summary>
    /// Intended to be applied to types and members marked as internal OR PRIVATE PROTECTED to indicate that friend access of ...

---
In reply to: [256221035](https://github.com/dotnet/machinelearning/pull/2526#discussion_r256221035) [](ancestors = 256221035)

Reply:
I can change "marked as internal" to "with internal scope," but I do not intend to push that change unless I have to update the PR anyway for some other reason. The root of your confusion, as far as I can tell, is not realizing that `private protected` implied the internal scope in its definition. I agree it's confusing, but the XML docs here are definitely *not* the place to resolve this confusion.

Reply:
In any case, merged https://github.com/dotnet/dotnet-api-docs/pull/1844 to clarify .NET docs.